File: frontend/lib/fovea/tracking.ts
import type { Point } from './engine.ts';

export type MotionStatus = 'moving' | 'stationary' | 'unknown';
export type TrackObject = {
  trackId: number;
  centroid: [number, number, number];
  min: [number, number, number];
  max: [number, number, number];
  dimensions: [number, number, number];
  points: number;
  velocity: [number, number, number] | null;
  speedMps: number | null;
  motionStatus: MotionStatus;
  motionFrame: 'world-compensated' | 'stationary-sensor-assumed' | 'sensor-relative-uncompensated';
  trackHistory: [number, number, number][];
  source: 'geometric-estimate' | 'ground-truth-annotation' | 'model-prediction';
  className?: string;
  trackingStatus: string;
};

export type TrackingConfig = { groundGrid: number; groundMargin: number; clusterSize: number; minClusterPoints: number; gateDistance: number; maxMissed: number; movingEnterMps: number; movingExitMps: number; stationarySensor: boolean };
export const TRACKING_DEFAULTS: TrackingConfig = { groundGrid: 1, groundMargin: .35, clusterSize: .9, minClusterPoints: 8, gateDistance: 3, maxMissed: 2, movingEnterMps: 1.2, movingExitMps: .7, stationarySensor: false };

type State = { id: number; centroid: [number, number, number]; velocity: [number, number, number]; timestamp: number; missed: number; history: [number, number, number][]; motionStatus: MotionStatus; object: TrackObject };
const dist = (a: number[], b: number[]) => Math.hypot(a[0] - b[0], a[1] - b[1], a[2] - b[2]);
const add = (a: number[], b: number[]) => [a[0] + b[0], a[1] + b[1], a[2] + b[2]] as [number, number, number];
const mul = (a: number[], s: number) => [a[0] * s, a[1] * s, a[2] * s] as [number, number, number];
const sub = (a: number[], b: number[]) => [a[0] - b[0], a[1] - b[1], a[2] - b[2]] as [number, number, number];

function separateGround(points: Point[], cfg: TrackingConfig) {
  const mins = new Map<string, number>();
  for (const p of points) { const k = `${Math.floor(p.x / cfg.groundGrid)}:${Math.floor(p.y / cfg.groundGrid)}`; mins.set(k, Math.min(mins.get(k) ?? p.z, p.z)); }
  return points.map(p => p.z <= (mins.get(`${Math.floor(p.x / cfg.groundGrid)}:${Math.floor(p.y / cfg.groundGrid)}`) ?? p.z) + cfg.groundMargin);
}

const TERRAIN_LABELS = new Set([
  0, 1, 3, 6, 7, 10, 23, 29, 30, 31, 32, 33, 34,
]);

const DYNAMIC_SEMANTIC_LABELS = new Set([8, 17]);

function classNameFromGeometry(dimensions: [number, number, number]): string {
  const [width, depth, height] = dimensions;
  const length = Math.max(width, depth);
  const thickness = Math.min(width, depth);

  if (height > 1.2 && height < 2.4 && length < 1.2) {
    return 'person';
  }

  if (height > 1.1 && height < 3.2 && length > 2.0 && length < 7.5) {
    return 'vehicle';
  }

  if (height > 1.0 && length > 3.0 && thickness < 0.8) {
    return 'wall/fence';
  }

  if (height > 2.0 && thickness < 1.0) {
    return 'pole/tree';
  }

  if (height < 1.2 && length > 1.0) {
    return 'barrier/object';
  }

  return 'unclassified obstacle';
}

export function extractObjects(points: Point[], semantic?: number[], names?: Record<number, string>, cfg: TrackingConfig = TRACKING_DEFAULTS): TrackObject[] {
  const ground = separateGround(points, cfg), idx: number[] = [];
  for (let i = 0; i < points.length; i++) {
    const terrain = semantic && TERRAIN_LABELS.has(semantic[i]);
    if (!ground[i] && !terrain) idx.push(i);
  }
  const bins = new Map<string, number[]>();
  for (const i of idx) { const p = points[i]; const key = `${Math.floor(p.x / cfg.clusterSize)}:${Math.floor(p.y / cfg.clusterSize)}:${Math.floor(p.z / cfg.clusterSize)}`; bins.set(key, [...(bins.get(key) || []), i]); }
  const out: TrackObject[] = [];
  while (bins.size && out.length < 80) {
    const first = bins.keys().next().value as string, queue = [first], ids = bins.get(first)!; bins.delete(first);
    while (queue.length) { const [x, y, z] = queue.shift()!.split(':').map(Number); for (let dx = -1; dx <= 1; dx++)for (let dy = -1; dy <= 1; dy++)for (let dz = -1; dz <= 1; dz++) { const k = `${x + dx}:${y + dy}:${z + dz}`, v = bins.get(k); if (v) { ids.push(...v); bins.delete(k); queue.push(k); } } }
    if (ids.length < cfg.minClusterPoints) continue;
    const ps = ids.map(i => points[i]);
    const xs = ps.map(p => p.x), ys = ps.map(p => p.y), zs = ps.map(p => p.z);
    const min = [Math.min(...xs), Math.min(...ys), Math.min(...zs)] as [number, number, number];
    const max = [Math.max(...xs), Math.max(...ys), Math.max(...zs)] as [number, number, number];
    const centroid = [xs.reduce((a, b) => a + b, 0) / ps.length, ys.reduce((a, b) => a + b, 0) / ps.length, zs.reduce((a, b) => a + b, 0) / ps.length] as [number, number, number];
    const obj: TrackObject = { trackId: 0, centroid, min, max, dimensions: sub(max, min), points: ps.length, velocity: null, speedMps: null, motionStatus: 'unknown', motionFrame: cfg.stationarySensor ? 'stationary-sensor-assumed' : 'sensor-relative-uncompensated', trackHistory: [centroid], source: semantic ? 'ground-truth-annotation' : 'geometric-estimate', trackingStatus: 'detected' };
    if (semantic) {
      const counts = new Map<number, number>();

      for (const i of ids) {
        counts.set(semantic[i], (counts.get(semantic[i]) || 0) + 1);
      }

      const best = [...counts.entries()].sort((a, b) => b[1] - a[1])[0]?.[0];

      if (best !== undefined) {
        const baseName = names?.[best] || `rellis-${best}`;
        obj.className = DYNAMIC_SEMANTIC_LABELS.has(best)
          ? `dynamic ${baseName}`
          : baseName;
      }
    } else {
      obj.className = classNameFromGeometry(obj.dimensions);
    }
    out.push(obj);
  }
  return out.sort((a, b) => b.points - a.points);
}

export class LocalTracker {
  private tracks = new Map<number, State>();
  private nextId = 1;
  private lastFrame: number | null = null;
  reset() { this.tracks.clear(); this.nextId = 1; this.lastFrame = null; }
  update(objects: TrackObject[], timestamp: number, frameIndex: number, cfg: TrackingConfig = TRACKING_DEFAULTS, jumped = false) {
    if (jumped || this.lastFrame !== null && frameIndex !== this.lastFrame + 1) this.reset();
    this.lastFrame = frameIndex;
    const speedValid = cfg.stationarySensor, mode: TrackObject['motionFrame'] = speedValid ? 'stationary-sensor-assumed' : 'sensor-relative-uncompensated';
    const pairs: { d: number; oi: number; id: number }[] = [];
    for (let oi = 0; oi < objects.length; oi++)for (const [id, t] of this.tracks) { const dt = Math.max(0, timestamp - t.timestamp), pred = add(t.centroid, mul(t.velocity, dt)), d = dist(objects[oi].centroid, pred); if (d <= cfg.gateDistance) pairs.push({ d, oi, id }); }
    pairs.sort((a, b) => a.d - b.d); const usedO = new Set<number>(), usedT = new Set<number>(), out: TrackObject[] = [];
    for (const p of pairs) { if (usedO.has(p.oi) || usedT.has(p.id)) continue; const prev = this.tracks.get(p.id)!, obj = objects[p.oi], dt = Math.max(1e-6, timestamp - prev.timestamp), vel = mul(sub(obj.centroid, prev.centroid), 1 / dt), speed = Math.hypot(...vel); let status: MotionStatus = 'unknown'; if (speedValid) status = prev.motionStatus === 'moving' ? (speed < cfg.movingExitMps ? 'stationary' : 'moving') : (speed > cfg.movingEnterMps ? 'moving' : 'stationary'); const history = [...prev.history.slice(-19), obj.centroid]; const tracked = { ...obj, trackId: p.id, velocity: speedValid ? vel : null, speedMps: speedValid ? speed : null, motionStatus: status, motionFrame: mode, trackHistory: history, trackingStatus: 'active' }; this.tracks.set(p.id, { id: p.id, centroid: obj.centroid, velocity: vel, timestamp, missed: 0, history, motionStatus: status, object: tracked }); out.push(tracked); usedO.add(p.oi); usedT.add(p.id); }
    for (let i = 0; i < objects.length; i++)if (!usedO.has(i)) { const id = this.nextId++, obj = { ...objects[i], trackId: id, motionFrame: mode, trackingStatus: 'active' }; this.tracks.set(id, { id, centroid: obj.centroid, velocity: [0, 0, 0], timestamp, missed: 0, history: [obj.centroid], motionStatus: 'unknown', object: obj }); out.push(obj); }
    for (const [id, t] of [...this.tracks]) if (!usedT.has(id) && !out.some(o => o.trackId === id)) { t.missed++; if (t.missed > cfg.maxMissed) this.tracks.delete(id); }
    return out.sort((a, b) => a.trackId - b.trackId);
  }
}

File: backend/foveamap/tracking.py
from collections import deque
from dataclasses import dataclass
import math
import numpy as np


@dataclass
class TrackingConfig:
    ground_grid: float = 1.0
    ground_margin: float = 0.35
    cluster_size: float = 0.9
    min_cluster_points: int = 8
    max_cluster_points: int = 20000
    max_detections: int = 80
    gate_distance: float = 3.0
    max_missed: int = 2
    moving_enter_mps: float = 1.2
    moving_exit_mps: float = 0.7
    stationary_sensor: bool = False


def _as_points(points):
    arr = np.asarray(points, dtype=float)
    if arr.ndim != 2 or arr.shape[1] < 3:
        raise ValueError('points must be an Nx3 or Nx4 array')
    if not np.isfinite(arr[:, :3]).all():
        raise ValueError('points contain non-finite coordinates')
    return arr[:, :3]


def separate_ground(points, config=TrackingConfig()):
    """Local-min height filter for uneven outdoor terrain.

    The lowest point in each XY grid cell is treated as local terrain. Points
    above that local minimum by ground_margin are candidate obstacles.
    """
    p = _as_points(points)
    if len(p) == 0:
        return np.zeros(0, dtype=bool)
    keys = np.floor(p[:, :2] / config.ground_grid).astype(int)
    mins = {}
    for key, z in zip(map(tuple, keys), p[:, 2]):
        mins[key] = min(mins.get(key, z), z)
    ground = np.array([z <= mins[tuple(key)] + config.ground_margin for key, z in zip(keys, p[:, 2])])
    return ground


def extract_objects(points, labels=None, semantic_names=None, config=TrackingConfig()):
    p = _as_points(points)
    if len(p) == 0:
        return []
    ground = separate_ground(p, config)
    candidate_idx = np.flatnonzero(~ground)
    if labels is not None:
        labels = np.asarray(labels)
        if len(labels) == len(p):
            # Ignore terrain-like labels when annotations are present, but do not
            # infer motion from class names.
            terrain = np.isin(labels, [0, 1, 3, 10, 23, 31, 33])
            candidate_idx = np.flatnonzero(~ground & ~terrain)
    if not len(candidate_idx):
        return []
    q = p[candidate_idx]
    bins = {}
    for local_i, key in enumerate(np.floor(q / config.cluster_size).astype(int)):
        bins.setdefault(tuple(key), []).append(local_i)
    objects = []
    while bins and len(objects) < config.max_detections:
        key = next(iter(bins))
        queue = deque([key])
        ids = bins.pop(key)
        while queue:
            base = queue.popleft()
            for dx in range(-1, 2):
                for dy in range(-1, 2):
                    for dz in range(-1, 2):
                        nb = (base[0] + dx, base[1] + dy, base[2] + dz)
                        if nb in bins:
                            ids.extend(bins.pop(nb))
                            queue.append(nb)
        if not config.min_cluster_points <= len(ids) <= config.max_cluster_points:
            continue
        original = candidate_idx[ids]
        cluster = p[original]
        mn, mx = cluster.min(0), cluster.max(0)
        obj = {
            'centroid': cluster.mean(0).tolist(),
            'min': mn.tolist(),
            'max': mx.tolist(),
            'dimensions': (mx - mn).tolist(),
            'points': int(len(cluster)),
            'source': 'geometric-estimate',
            'label': None,
        }
        if labels is not None and len(labels) == len(p):
            vals, counts = np.unique(labels[original], return_counts=True)
            best = int(vals[np.argmax(counts)])
            obj['label'] = best
            if semantic_names:
                obj['className'] = semantic_names.get(best, f'rellis-{best}')
                obj['source'] = 'ground-truth-annotation'
        objects.append(obj)
    objects.sort(key=lambda x: (-x['points'], x['centroid'][0], x['centroid'][1]))
    return objects


def clusters(points, labels=None, size=.8, minimum=8):
    """Compatibility wrapper used by the original CLI/tests."""
    cfg = TrackingConfig(cluster_size=size, min_cluster_points=minimum, ground_margin=-math.inf)
    if labels is not None:
        labels = np.asarray(labels)
        p = np.asarray(points)[labels == 3, :3] if len(labels) == len(points) else np.asarray(points)[:, :3]
    else:
        p = np.asarray(points)[:, :3]
    return extract_objects(p, config=cfg)


class Tracker:
    """Pose-aware gated one-to-one centroid tracker.

    With valid world_from_lidar poses, matching and velocity are computed in a
    common world frame. Without poses, velocity is reported as sensor-relative
    and uncompensated unless stationary_sensor=True.
    """
    def __init__(self, max_distance=3.0, ttl=1.0, config=None):
        self.config = config or TrackingConfig(gate_distance=max_distance, max_missed=max(1, int(math.ceil(ttl))))
        self.next_id = 1
        self.tracks = {}
        self.last_frame = None

    def reset(self):
        self.next_id = 1
        self.tracks = {}
        self.last_frame = None

    def _frame_point(self, centroid, pose):
        c = np.asarray([*centroid, 1.0], dtype=float)
        return (pose @ c)[:3] if pose is not None else c[:3]

    def update(self, objects, timestamp, frame_index=None, pose=None, reset=False):
        if reset or (frame_index is not None and self.last_frame is not None and frame_index != self.last_frame + 1):
            self.reset()
        self.last_frame = frame_index
        if pose is not None:
            pose = np.asarray(pose, dtype=float)
            if pose.shape != (4, 4) or not np.isfinite(pose).all():
                raise ValueError('pose must be a finite 4x4 world_from_lidar matrix')
        if not self.config.stationary_sensor and pose is None:
            frame_mode = 'sensor-relative-uncompensated'
            speed_valid = False
        elif pose is None:
            frame_mode = 'stationary-sensor-assumed'
            speed_valid = True
        else:
            frame_mode = 'world-compensated'
            speed_valid = True

        detections = []
        for obj in objects:
            world = self._frame_point(obj['centroid'], pose)
            detections.append((obj, world))

        pairs = []
        for di, (_, centre) in enumerate(detections):
            for tid, tr in self.tracks.items():
                dt = max(0.0, float(timestamp) - tr['timestamp'])
                pred = tr['world'] + tr['velocity'] * dt
                d = float(np.linalg.norm(centre - pred))
                if d <= self.config.gate_distance:
                    pairs.append((d, di, tid))
        pairs.sort()
        used_d, used_t, outputs = set(), set(), []
        for _, di, tid in pairs:
            if di in used_d or tid in used_t:
                continue
            obj, centre = detections[di]
            tr = self.tracks[tid]
            dt = max(1e-6, float(timestamp) - tr['timestamp'])
            velocity = (centre - tr['world']) / dt
            status = tr.get('motionStatus', 'unknown')
            speed = float(np.linalg.norm(velocity))
            if speed_valid:
                if status == 'moving':
                    status = 'stationary' if speed < self.config.moving_exit_mps else 'moving'
                else:
                    status = 'moving' if speed > self.config.moving_enter_mps else 'stationary'
            else:
                status = 'unknown'
            history = [*tr['history'][-19:], obj['centroid']]
            state = {'id': tid, 'world': centre, 'velocity': velocity, 'timestamp': float(timestamp), 'missed': 0, 'history': history, 'motionStatus': status}
            self.tracks[tid] = state
            outputs.append(self._decorate(obj, state, frame_mode, speed_valid))
            used_d.add(di); used_t.add(tid)

        for di, (obj, centre) in enumerate(detections):
            if di in used_d:
                continue
            tid = self.next_id; self.next_id += 1
            state = {'id': tid, 'world': centre, 'velocity': np.zeros(3), 'timestamp': float(timestamp), 'missed': 0, 'history': [obj['centroid']], 'motionStatus': 'unknown'}
            self.tracks[tid] = state
            outputs.append(self._decorate(obj, state, frame_mode, speed_valid))

        for tid in list(self.tracks):
            if tid not in used_t and all(o['trackId'] != tid for o in outputs):
                self.tracks[tid]['missed'] += 1
                if self.tracks[tid]['missed'] > self.config.max_missed:
                    del self.tracks[tid]
        outputs.sort(key=lambda x: x['trackId'])
        return outputs

    def _decorate(self, obj, state, frame_mode, speed_valid):
        v = state['velocity']
        speed = float(np.linalg.norm(v)) if speed_valid else None
        return {
            **obj,
            'trackId': state['id'],
            'velocity': v.tolist() if speed_valid else None,
            'speedMps': speed,
            'motionStatus': state['motionStatus'] if speed_valid else 'unknown',
            'motionFrame': frame_mode,
            'trackHistory': state['history'],
            'trackingStatus': 'active',
        }
